src/csharp/Grpc.Core.Api/ServerServiceDefinition.cs

```diff
@@ -18,33 +18,31 @@
 
 using System;
 using System.Collections.Generic;
-using System.Collections.ObjectModel;
-using System.Linq;
-using Grpc.Core.Interceptors;
-using Grpc.Core.Internal;
-using Grpc.Core.Utils;
 
 namespace Grpc.Core
 {
     /// <summary>
-    /// Mapping of method names to server call handlers.
+    /// Stores mapping of methods to server call handlers.
     /// Normally, the <c>ServerServiceDefinition</c> objects will be created by the <c>BindService</c> factory method 
     /// that is part of the autogenerated code for a protocol buffers service definition.
     /// </summary>
     public class ServerServiceDefinition
     {
-        readonly ReadOnlyDictionary<string, IServerCallHandler> callHandlers;
+        readonly IReadOnlyList<Action<ServiceBinderBase>> addMethodActions;
 
-        internal ServerServiceDefinition(Dictionary<string, IServerCallHandler> callHandlers)
+        internal ServerServiceDefinition(List<Action<ServiceBinderBase>> addMethodActions)
         {
-            this.callHandlers = new ReadOnlyDictionary<string, IServerCallHandler>(callHandlers);
+            this.addMethodActions = addMethodActions.AsReadOnly();
         }
 
-        internal IDictionary<string, IServerCallHandler> CallHandlers
+        /// <summary>
+        /// Forwards all the previously stored <c>AddMethod</c> calls to the service binder.
+        /// </summary>
+        internal void BindService(ServiceBinderBase serviceBinder)
```

Comment:
We could even make this method public and thus allow ServerServiceDefinition to be used by other grpc implementations than Grpc.Core.
Potentially this would eliminate the need for having two BindService overloads in the generated code.

Reply:
This sounds interesting and I would definitely prefer reducing duplication. I have a couple of questions though I think I already know the answer.

1. Why keep the Builder class? Seems like `ServerServiceDefinition` is just a container for a set of actions to be taken during `BindService`. In which case it's simpler to just add the `AddMethod` methods to the `ServerServiceDefinition` class itself. This will require a bit of tweeking of the generated code but otherwise feasible. I'm assuming  back compat must be maintained since the Builder is public and therefore already exposed?

2. The naming makes the usage a little odd. What you end up with is needing to invoke `Foo.BindService(new FooImpl()).BindService(new ServiceBinderImpl())` The chain of `.BindService` looks odd and unintuitive but it does work. Logically, it should look like `Foo.CollectServices(new FooImpl()).BindServices(new ServiceBinderImpl())`. I'm assuming we'd want to avoid it because it would require regenerating the code.

Reply:
1. yes, keeping the Builder is for backwards compatibility - we don't want to force the entire world to regenerate just because of this change (also regenerating might not be easy for some - they might be depending on third party nuget packages that already contain pre-compiled generated .cs files).

2. we can rename the ServerServiceDefinition.BindService because it's newly added in this PR, but I'm afraid the name of "BindService" method in the generated code must stay unchanged.

Reply:
Sounds good.